ftl/common/builder.py

```diff
@@ -55,34 +63,105 @@ class JustApp(Base):
     def __init__(self, ctx):
         super(JustApp, self).__init__(ctx)
 
-    def CreatePackageBase(self, base_image):
+    def Build(self):
         """Override."""
         # JustApp doesn't install anything, it just appends
         # the application layer, so return the base image as
         # our package base.
-        return base_image
+        return
+
+    class AppLayer(build_layer.CacheLayerImage):
```

Comment:
How will this layer be cached? Maybe the sha of the context? Or maybe we just skip that part.

Reply:
it isn't cached, this is a misnomer currently as I made the base abstract class require cache methods.  the app layer isn't cached so deriving from CachLayerImage is confusing, I have a note to change it

Reply:
currently the AppLayer is always built for the reasons we did it prior.  i can change it to be treated like the other layers, I wasn't sure what the key should be though.  sha of the context makes sense

Reply:
it might make sense to cache it like the other layers

Reply:
when you say 'sha of the context' what are you referring?  what would be passed to hashlib,sha256()?

Reply:
> when you say 'sha of the context' what are you referring? what would be passed to hashlib,sha256()?

You would probably need to add a method to the context interface - it would roughly be a concatenation of all files in the context in a stable order. I'm not sure if it would be worth it, since the context caching should be somewhat automated during the upload itself. As long as the layer always has the same digest, we get some built-in caching.

Reply:
I'm just going to use the 'built-in' caching we have been using for now